### DISCUSSION TITLE
Concurent Continuous Delivery behaviour

### DISCUSSION THREAD

**Discussion by sbancal** [EPFL-ENAC/epfl-enac-deploy-action#11] [Polls]
Related to https://github.com/EPFL-ENAC/enac-cd-app/issues/9 ,

We have to decide the behaviour expected when 2 commit reach the Github repo faster than the _continuous deployment_ has deployed the 1st one.

One important thing is to output the SHA1 in the output so that when we come back to an action's output we are sure of what was deployed. This will be addressed by https://github.com/EPFL-ENAC/enacit-ansible/issues/127

**Scenario A)** would be as asked by https://github.com/EPFL-ENAC/enac-cd-app/issues/9 : A newer action will simply be attached to an already running process.
👍🏼  this is already implemented
👎🏼  there is no guarantee of the deployed app version, even with a ✔️ , the most probably deployed previous is the 1st one.

**Scenario B)** A newer commit would interrupt an unfinished continuous deployment that is no longer needed.
👍🏼  be sure that a ✔️ on latest commit means that this version is deployed
👎🏼  what if the interruption comes when a sensitive step is run (like potentially a DB migration)?

**Scenario C)** A newer commit would wait for previous continuous deployment to finish before running itself. If several newer commits occurs before CD has finished, then only the latest will be launched. The other will be aborted.
👍🏼  we're sure of the consistency
👎🏼  that means longer deployment waiting time for the end user

**Scenario D)** Something else

Update 2023-05-15 1PM, considering the opinions expressed @guilbep @evanjt 

**Scenario E)** Knowing that _app-deploy_ will always deploy latest commit.
If the a second request comes before `git pull` operation is run, then new request will be aggregated with the previous request.
If it comes after the `git pull`, then it'll be queued like described in **C**.

As we are also discussing about the interruption of a process.
I'd suggest that without any admin intervention the process is not interrupted.
If someone asks to cancel an action via Github, then it'll be a hard (and easy) `kill`, as if the sysadmin were doing <ctrl-c> in their terminal while Ansible was running.

**Comment by guilbep** [comment]
- We could also; keep the current behavior and add a way to intercept the CTRL+C in the github actions to call a 'cancel' to the enac-cd-app.

That way if. a user 'cancel' a workflow; it will properly cancel the deployment. And if we restart a workflow it will behave properly.

See documentation here https://docs.github.com/en/actions/managing-workflow-runs/canceling-a-workflow

**Reply by sbancal** [reply]
This would be a new feature ... and would be unsafe for the 👎🏼 scenario described in **B**
(by the way ... I'm not convinced that 👎🏼 scenario must be addressed)

**Reply by guilbep** [reply]
I'd prefer this option D)

keep the current behavior and add a way to intercept the CTRL+C in the github actions to call a 'cancel' to the enac-cd-app.
That way if. a user 'cancel' a workflow; it will properly cancel the deployment. And if we restart a workflow it will behave properly.

That way we can force a redeploy of any commit if we wish. Option A) does not seem a good option anymore

**Reply by sbancal** [reply]
There is no way to deploy a specific commit. It's always the latest of the chosen branch.

If you want to deploy a different commit to the latest one, then you must drop some commits (ugly) ... of `git revert` them.

**Reply by sbancal** [reply]
also «properly cancel the deployment» needs more details on how we do that.
1. If that means we interrupt the Ansible process, that's quite feasible.
2. If that means we ask the process run by Ansible to cleanly stop, then that means we have to implement something in every project. This is not an option in my eyes.
3. If that means we let Ansible finish current step, but stop before next step ... that would mean quite much of rewrite.

**Reply by evanjt** [reply]
I think as long as the behaviour is known and documented, I am happy with E because in any case, we have to adapt to a scenario we put ourselves (as developers) in and this seems like an appropriate compromise.

**Comment by evanjt** [comment]
I think my intuition would be C, with the ability to cancel any previous run, failing it with a X and having the next CD in line spawn.
